### PR TITLE
Issue #2511252 by cbanman: Create a service for getting enabled produ…

### PR DESCRIPTION
**Pull request by crbanman**
…ct variations

### CODE REVIEW

File: modules/product/src/ProductVariationManager.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  protected $variationStorage;
+
+  /**
+   * The event dispatcher.
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $eventDispatcher;
+
+
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: Accidental double newline.

File: modules/product/src/ProductVariationManager.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    $variations = $product->variations;
+    $enabledVariations = [];
+
+    $query = \Drupal::entityQuery('commerce_product_variation')
+    ->condition('status', TRUE);
+    $group = $query->orConditionGroup();
+
+    foreach($variations as $variation) {
+      $id = $variation->getValue();
+      $group->condition('variation_id', $id);
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: You can just do $query->condition('variation_id', $ids); and EQ will make it an IN condition.

File: modules/product/src/ProductVariationManager.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getEnabledVariations(Product $product) {
+    $variations = $product->variations;
+    $enabledVariations = [];
+
+    $query = \Drupal::entityQuery('commerce_product_variation')
+    ->condition('status', TRUE);
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: Missing indentation.

File: modules/product/src/ProductVariationManager.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    $group = $query->orConditionGroup();
+
+    foreach($variations as $variation) {
+      $id = $variation->getValue();
+      $group->condition('variation_id', $id);
+    }
+
+    $result = $query->condition($group)->execute();
+
+    if(!empty($result)){
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: if $result is empty we should exit before firing the event, just return [];

File: modules/product/src/ProductVariationManagerInterface.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+namespace Drupal\commerce_product;
+use Drupal\commerce_product\Entity\Product;
+
+/**
+ * Defines the interface for product variation managers.
+ */
+interface ProductVariationManagerInterface {
+
+  /**
+   * Gets an array containing all the enabled variations of the given product.
```
Comment (bojanz) [inline modules/product/src/ProductVariationManagerInterface.php]: "Gets a list of enabled variations for the given product."

File: modules/product/src/ProductVariationManagerInterface.php [hunk 6]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\commerce_product\ProductVariationManagerInterface.
+ */
+
+namespace Drupal\commerce_product;
+use Drupal\commerce_product\Entity\Product;
```
Comment (bojanz) [inline modules/product/src/ProductVariationManagerInterface.php]: Needs newline between namespace and use statement.

File: modules/product/src/ProductVariationManagerInterface.php [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * Defines the interface for product variation managers.
+ */
+interface ProductVariationManagerInterface {
+
+  /**
+   * Gets an array containing all the enabled variations of the given product.
+   *
+   * @param \Drupal\commerce_product\Entity\Product $product
```
Comment (bojanz) [inline modules/product/src/ProductVariationManagerInterface.php]: Needs description for the $product param (on the next line, indented, "The product.")

File: modules/product/src/Event/AvailableVariationsEvent.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Drupal\commerce_product\Event;
+
+use Drupal\commerce_product\Entity\Product;
+use Symfony\Component\EventDispatcher\Event;
+
+class AvailableVariationsEvent extends Event {
+
+  /**
+   * The product containing variations to be filtered.
```
Comment (bojanz) [inline modules/product/src/Event/AvailableVariationsEvent.php]: "The parent product." is enough.

File: modules/product/src/Event/AvailableVariationsEvent.php [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+   *
+   * @var \Drupal\commerce_product\Entity\Product
```
Comment (bojanz) [inline modules/product/src/Event/AvailableVariationsEvent.php]: Let's use \Drupal\commerce_product\ProductInterface, always better than mentioning a specific class. Same in the constructor docblock and typehint.

**Comment by bojanz** [comment]
There are still mentions of FilterVariationsEvent, the rename was incomplete. Check the class file docblock, the ProductEvents constant and id.

**Comment by crbanman** [comment]
Made changes as per review.

File: modules/product/src/Event/AvailableVariationsEvent.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+   * An array of enabled variations.
+   *
+   * @var array
+   */
+  protected $variations;
+
+  /**
+   * Constructs a new AvailableVariationsEvent object.
+   *
+   * @param \Drupal\commerce_product\ProductInterface
```
Comment (bojanz) [inline modules/product/src/Event/AvailableVariationsEvent.php]: These two @param docblocks are incomplete (no variable name for product, no descriptions for both)

File: modules/product/src/ProductVariationManager.php [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    $this->eventDispatcher = $eventDispatcher;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getEnabledVariations(ProductInterface $product) {
+    $variations = $product->variations;
+    $ids = [];
+
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: Let's kill this newline.

File: modules/product/src/ProductVariationManager.php [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    foreach($variations as $variation) {
+      $ids[] = $variation->getValue()['target_id'];
+    }
+
+    $query = \Drupal::entityQuery('commerce_product_variation')
+      ->condition('status', TRUE)
+      ->condition('variation_id', $ids, "IN");
+
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: And this one

File: modules/product/src/ProductVariationManager.php [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    $result = $query->execute();
+
+    if(empty($result)){
+      return [];
+    }
+
+    $enabledVariations = $this->variationStorage->loadMultiple($result);
+
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: And this one.

**Comment by crbanman** [comment]
Updated.

File: modules/product/src/ProductVariationManagerInterface.php [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * Defines the interface for product variation managers.
+ */
+interface ProductVariationManagerInterface {
+
+  /**
+   * Gets a list of enabled variations for the given product.
+   *
+   * @param \Drupal\commerce_product\ProductInterface $product
+   *   The product.
```
Comment (bojanz) [inline modules/product/src/ProductVariationManagerInterface.php]: You need a newline before @return

File: modules/product/src/ProductVariationManager.php [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+   * The event dispatcher.
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $eventDispatcher;
+
+  /**
+   * Constructs a ProductVariationManager object.
+   *
+   * @param \Drupal\Core\Entity\EntityManagerInterface $entityManager
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: These two @param have no description in the next line.

File: modules/product/src/ProductVariationManager.php [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  public function __construct(EntityManagerInterface $entityManager, EventDispatcherInterface $eventDispatcher) {
+    $this->variationStorage = $entityManager->getStorage('commerce_product_variation');
+    $this->eventDispatcher = $eventDispatcher;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getEnabledVariations(ProductInterface $product) {
+    $variations = $product->variations;
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: No need to declare $variations, you aren't going to use them again, just do foreach ($product->variations as $item) {

File: modules/product/src/ProductVariationManager.php [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    $ids = [];
+    foreach($variations as $variation) {
+      $ids[] = $variation->getValue()['target_id'];
```
Comment (bojanz) [inline modules/product/src/ProductVariationManager.php]: can be just $ids[] = $item->target_id;

**Comment by crbanman** [comment]
Fixed. I also realized there was a ; at the end of the arguments in commerce_product.services.yml which would messed with everything.

File: modules/product/src/Event/ProductEvents.php [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+namespace Drupal\commerce_product\Event;
+
+final class ProductEvents {
+
+  /**
+   * Name of the event fired when filtering variations.
+   *
+   * @Event
+   *
+   * @see \Drupal\commerce_product\Event\AvailableVariationsEvent
```
Comment (bojanz) [inline modules/product/src/Event/ProductEvents.php]: Whoops :) That @see needs to be updated

**Pull request merged**